oarepo_vocabularies/ui/theme/assets/semantic-ui/js/oarepo_vocabularies_ui/form/components/VocabularyTreeSelectField/VocabularyTreeSelectField.jsx

```diff
@@ -0,0 +1,526 @@
+        handleArrowUp(e, index, data);
+        break;
+
+      case "ArrowDown":
+        handleArrowDown(e, index, data);
+        break;
+
+      case "ArrowLeft":
+        handleArrowLeft(index);
+        break;
+
+      case "ArrowRight":
+        handleArrowRight(index);
+        break;
+
+      case "Enter":
+      case " ":
+        handleEnterSpace(e, index, data);
+        break;
+    }
+  };
+
+  const renderColumn = (column, index) => {
+    return (
+      <Grid.Column key={index} className="tree-column">
+        {column.map((option, i) => {
+          if (
+            index == 0 ||
+            option.hierarchy.ancestors[0] == parentsState[index - 1]
+          ) {
+            return (
+              <Grid.Row
+                key={option.value}
+                className={
+                  option.value == parentsState[index] ? "open spaced" : "spaced"
+                }
+              >
+                {multiple && (
+                  <Checkbox
+                    checked={
+                      selectedState.findIndex(
+                        (item) => item.value === option.value
+                      ) !== -1
+                    }
+                    indeterminate={selectedState.some((item) =>
+                      item.hierarchy.ancestors.includes(option.value)
+                    )}
+                    onChange={(e) => {
+                      handleSelect(option, e);
+                    }}
+                  />
+                )}
+                <Button
+                  basic
+                  color="black"
+                  onClick={openHierarchyNode(option.value, index)}
+                  onDoubleClick={(e) => {
+                    handleSelect(option, e);
+                  }}
+                  onKeyDown={(e) => {
+                    handleKey(e, index);
+                  }}
+                  tabIndex={0}
+                >
+                  {option.hierarchy.title[0]}
+                </Button>
+                {option.element_type == "parent" && (
+                  <Button onClick={openHierarchyNode(option.value, index)}>
+                    {index !== columnsCount - 1 && (
+                      <Icon name="angle right" color="black" />
+                    )}
+                  </Button>
+                )}
+              </Grid.Row>
+            );
+          }
+        })}
+      </Grid.Column>
+    );
+  };
+
+  return (
+    <React.Fragment>
+      <SelectField
+        selectOnBlur={false}
+        optimized={optimized}
+        onBlur={() => setFieldTouched(fieldPath)}
+        deburr
+        fieldPath={fieldPath}
+        multiple={multiple}
+        featured={featuredOptions}
+        options={serializedOptions}
+        onClick={(e) => handleOpen(e)}
+        onChange={handleChange}
+        value={multiple ? value.map((o) => o?.id) : value?.id}
+        {...uiProps}
+      />
+      <label className="helptext">{helpText}</label>
+
+      {openState && (
+        <Modal
```

Comment:
I would extract Modal to its own component (`VocabularyTreeSelectModal`) to simplify this component